pkg/transport/client_low.go

```diff
@@ -0,0 +1,165 @@
+
+	logger        logrus.FieldLogger
+	compress      bool
+	customHeaders map[string]string
+	debugBody     bool
+	enableMetrics bool
+	requestSem    util.Semaphore
+	userAgent     string
+	backoff       util.BackoffFactory
+
+	Client *http.Client
+}
+
+func sendMetricIfChanged(statser stats.Statser, metricName string, tags gostatsd.Tags, m *metricTracker) {
+	v := atomic.LoadUint64(&m.cur)
+	if v != m.prev {
+		m.prev = v
+		m.pending = 3 // send this metric for the next 3 intervals
+	}
+	if m.pending > 0 {
+		m.pending--
+		statser.Gauge(metricName, float64(v), tags)
+	}
+}
+
+func (hc *Client) emitMetrics(statser stats.Statser) {
+	if !hc.enableMetrics {
+		return
+	}
+	sendMetricIfChanged(statser, "transport.fail", gostatsd.Tags{"failure:send"}, &hc.messagesDropped)
+	sendMetricIfChanged(statser, "transport.fail", gostatsd.Tags{"failure:compress"}, &hc.messagesFailCompress)
+	sendMetricIfChanged(statser, "transport.fail", gostatsd.Tags{"failure:marshal"}, &hc.messagesFailMarshal)
+	sendMetricIfChanged(statser, "transport.retried", nil, &hc.messagesRetried)
+	sendMetricIfChanged(statser, "transport.sent", nil, &hc.messagesSent)
+	sendMetricIfChanged(statser, "transport.fail", gostatsd.Tags{"failure:mutex"}, &hc.messagesTimedOut)
+}
+
+// PostRaw will start a goroutine (semaphore allowing) which attempts to POST the provided data
+// to the provided URL.  It is fire-and-forget by design.
+func (hc *Client) PostRaw(ctx context.Context, url, contentType, encoding string, headers map[string]string, body []byte) {
+	if !hc.requestSem.Acquire(ctx) {
+		atomic.AddUint64(&hc.messagesTimedOut.cur, 1)
+		return
+	}
+
+	go func() {
+		defer hc.requestSem.Release()
+		hc.do(ctx, url, contentType, encoding, body, headers)
+	}()
+}
+
+func (hc *Client) do(ctx context.Context, url, contentType, encoding string, body []byte, headers map[string]string) {
+	var err error
+
+	bo := hc.backoff()
+	doPost := hc.constructPost(ctx, body, url, contentType, encoding, headers)
+
+	for {
+		if err = doPost(); err == nil {
+			atomic.AddUint64(&hc.messagesSent.cur, 1)
+			return
+		}
+
+		next := bo.NextBackOff()
+		if next == backoff.Stop {
+			break
+		}
+
+		hc.logger.WithError(err).Warn("failed to send, retrying")
+
+		if !interruptableSleep(ctx, next) {
+			break
+		}
+
+		atomic.AddUint64(&hc.messagesRetried.cur, 1)
+	}
+
+	atomic.AddUint64(&hc.messagesDropped.cur, 1)
+	hc.logger.WithError(err).Info("failed to send, giving up")
+}
+
+func (hc *Client) constructPost(ctx context.Context, body []byte, url, contentType, encoding string, headers map[string]string) func() error {
+	return func() error {
+		req, err := http.NewRequest("POST", url, bytes.NewReader(body))
+		if err != nil {
+			return fmt.Errorf("unable to create http.Request: %v", err)
+		}
+
+		req = req.WithContext(ctx)
+
+		// Base headers
+		req.Header.Set("Content-Type", contentType)
+		req.Header.Set("Content-Encoding", encoding)
```

Comment:
Encoding and type could be enums for a little bit more type safety. Currently there is a risk to pass two strings in swapped positions.

Reply:
Swapping is definitely a hazard, but I don't know how to prevent it.  Go lacks enums (near as I can tell), and will happily promote a string to an alias of it, breaking the type safety: https://play.golang.org/p/lzAsJ8gGB_O

I could do a sanity check for valid values, but that seems overkill.  Is there something I'm missing?

Reply:
Yeah, what we are seeing in that example is the behavior of "untyped literals". This does not compile though because `x` is typed vs untyped literal. https://play.golang.org/p/wiCUwuQqj5g

Reply:
Ahh! I see now.  https://play.golang.org/p/xItV45OW49V

I'll fix it up in the cleanup commit (waiting on other feedback, I know the team is busy atm)

- [ ] this

Reply:
https://blog.golang.org/constants see the `String constants` section